iroha_network/src/lib.rs

```diff
@@ -0,0 +1,310 @@
+use std::{
+    convert::{TryFrom, TryInto},
+    error::Error,
+    io::prelude::*,
+    net::{IpAddr, SocketAddr, TcpListener, TcpStream},
+    str::FromStr,
+};
+
+pub mod prelude {
+    //! Re-exports important traits and types. Meant to be glob imported when using `iroha_network`.
+
+    #[doc(inline)]
+    pub use crate::{Request, Response, Stream};
+}
+
+pub const BUFFER_SIZE: usize = 2048;
+
+pub struct Network {
+    server_connection: TcpStream,
+}
+
+impl Network {
+    /// Creates a new client and connects it to the server.
+    pub fn connect(server_url: &str) -> Result<Network, Box<dyn Error>> {
+        Ok(Network {
+            server_connection: TcpStream::connect(server_url)?,
+        })
+    }
+
+    /// Sends `request` over established and persistent connection and gets `Response`.
+    pub fn send_request(&mut self, request: Request) -> Result<Response, Box<dyn Error>> {
+        Self::send_over_stream(&mut self.server_connection, request)
+    }
+
+    /// Establishes connection, sends `request` closes connection and returns `Response`.
+    pub fn send_request_to(server_url: &str, request: Request) -> Result<Response, Box<dyn Error>> {
+        let mut stream = TcpStream::connect(server_url)?;
+        Self::send_over_stream(&mut stream, request)
+    }
+
+    /// Listens on localhost on port `port`.
+    /// When there is an incoming connections passes it's `Stream` to `handler`.
+    pub fn listen<H>(port: u16, mut handler: H) -> Result<(), Box<dyn Error>>
+    where
+        H: FnMut(Box<dyn Stream>) -> Result<(), Box<dyn Error>>,
+    {
+        let listener = TcpListener::bind(SocketAddr::new(IpAddr::from_str("127.0.0.1")?, port))?;
+        while let Some(stream) = listener.incoming().next() {
+            handler(Box::new(stream?))?;
+        }
+        Ok(())
+    }
+
+    /// Helper function to call inside `listen` `handler` function to parse and send response.
+    /// The `handler` specified here will need to generate `Response` from `Request`.
+    pub fn handle_message<H>(mut stream: impl Stream, mut handler: H) -> Result<(), Box<dyn Error>>
+    where
+        H: FnMut(Request) -> Result<Response, Box<dyn Error>>,
+    {
+        let mut buffer = [0u8; BUFFER_SIZE];
+        let read_size = stream.read(&mut buffer).expect("Request read failed.");
```

Comment:
Maybe we can read_to_vec instead of preallocated buffer? Some Transactions could be larger than BUFFER_SIZE

Reply:
I think you mean `read_to_end`. `read_to_end` from what I know will only complete when the stream is dropped by either client or server. But client can't drop the stream as it waits fro server response, therefore server call to `read_to_end` will never complete.

So to sum up, I suggest we stay with the use of buffer for now.